Test Failure: io.openliberty.microprofile.health.file.healthcheck.internal_fat timed out (full mode)

riyaaroraa opened the pull request:
fixes #32637

Changes Made

- Split the original FAT suite into three focused test classes for clarity and isolation:

1. SimpleFileBasedHealthCheckTest — main functional coverage of file creation and readiness/liveness behaviour
2. LongStartupHealthCheckTest — extended startup validation scenarios
3. LongCheckIntervalHealthCheckTest — periodic update interval and non-thrashing checks

- Introduced a new utility class HealthFileUtils to centralize file polling, timing, and validation logic (with OS-aware timeout adjustments).
- Added stabilization logic (stabilize() helper) in LongCheckIntervalHealthCheckTest to wait for scheduler ticks before sampling timestamps.
- Adjusted timing windows to prevent false negatives in slower environments (EE11)
- Improved server start/stop handling and cleanup between test runs to prevent overlapping servers or missing logs.

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=fc266603-ca72-4292-a9bd-6e6be1bfe706) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=91e4da77-859e-4472-89ff-219527842a3b) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+    }
+
+    /** Wait for a file to contain a specific substring. */
+    public static boolean waitForFileContains(Path p, String needle, long timeoutSeconds, long pollMillis) {
```
Not used?
But also in general, Liberty test infra contains utility method that covers this.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+        }
+    }
+
+    public static boolean isWindows() {
```
Simplicity bundle should contian method for this already.
https://github.com/OpenLiberty/open-liberty/blob/integration/dev/fattest.simplicity/src/com/ibm/websphere/simplicity/Machine.java

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/FATSuite.java:
```diff
@@ -58,14 +59,13 @@ public static boolean isAllHealthCheckFilesCreated(File serverRootDirFile, int r
             }
 
             Log.info(FATSuite.class, methodName,
-                     String.format("At attempt %d with the following: HealthDir[%s], startedFile[%s], liveFile[%s], readyFile[%s].Max attempt is: %d.", attemptNumber, isDirExist,
```
Seems like these utility methods have been replaced. Can just delete it then.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+                if (baselineMillis >= 0 && lm > baselineMillis) {
+                    return false; // it updated
+                }
+                if (baselineMillis < 0 && lm > 0) {
```
It's good to check, but we'd never get in this scenario? We're already in the running phase of the test where we've already validated files exist by this point.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+        long window = Duration.ofSeconds(adjustSecondsForWindows(windowSeconds)).toNanos();
+        long poll = Math.max(25L, pollMillis);
+        while (System.nanoTime() - start < window) {
+            if (f != null && f.exists()) {
```
It's good to check, but we'd never get in this scenario? We're already in the running phase of the test where we've already validated files exist by this point.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+        return os != null && os.toLowerCase().contains("win");
+    }
+
+    /** Apply a small cushion to timeouts on Windows runners (file I/O/locking tends to be slower). */
```
We don't need a larger time windows for Windows machines.

The issue with platform timing is that there is a mismatch between when the FAT test starts running and when the server starts running (with the app). 

Even with IO delay, it wouldn't be require 50% increase in timeout window.

Ideally, they start at the same time and everything lines up.
But sometimes, Server start at  time X, and fat test starts at time X + Y  or the Fat test starts before the server starts. Which results in timing issues.
This affects windows and z/os / os/400 machines.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/LongStartupHealthCheckTest.java:
```diff
@@ -0,0 +1,123 @@
+        Assert.assertFalse(Constants.LIVE_SHOULD_NOT_HAVE_CREATED, HealthFileUtils.getLiveFile(root).exists());
+
+        Assert.assertTrue(Constants.STARTED_SHOULD_NOT_HAVE_CREATED,
+                          HealthFileUtils.waitForFileMissing(HealthFileUtils.getStartFile(root), 8, 200));
```
waitForFileMissing just checks that the file doesn't exist and exits the check. This isn't what we're hoping for?
This means that this block and the next block just check withing a very close amount of time that the files aren't created.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/LongStartupHealthCheckTest.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @Test
+    public void StartedHealthCheckTestLongStartupInterval() throws Exception {
```
Overall, we're going to still probably hit timing issues.
Better solution would be to use the tests to check the traces/logs.
That way we have a better verification of whether the startupCheckInterval is working/abiding by the configured value.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/LongCheckIntervalHealthCheckTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    @Test
+    public void HealthCheckTestLongCheckInterval() throws Exception {
```
Overall, we're going to still probably hit timing issues.
Better solution would be to use the tests to check the traces/logs.
That way we have a better verification of whether the checkInterval is working/abiding by the configured value.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/LongCheckIntervalHealthCheckTest.java:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    /** sample lastModified twice until it stops moving (short cap) */
```
Don't need Nano precision , we're waiting in millis anyways. Can avoid perf of nanos.

This method at best just checks if the file doesn't update within 500 ms. And assumes it may keep updating constantly for 6 seconds? (why 6?).

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/BaseHealthFilesTest.java:
```diff
@@ -0,0 +1,101 @@
+     * Proactively hit the health endpoints once the server is up to
+     * force mpHealth to initialize and write the /health directory and files.
+     */
+    protected static void triggerHealthEndpoints(LibertyServer server) throws Exception {
```
Triggering health endpoints do not trigger creation of the File based health checks.
The File based health checks are automatically created independent of any requests through the REST API.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/BaseHealthFilesTest.java:
```diff
@@ -0,0 +1,101 @@
+    protected static final String SERVER_LONG_CHECK_INTERVAL = "HealthServerLongCheckInterval";
+
+    // Wait settings: up to 3 minutes, 1s polling
+    private static final int FILE_WAIT_TRIES = 180;
```
The parameter for method this is sent into is the total time window, might reconsider renaming. Tries sounds misleading.

But also, waiting for 3 minutes for the files to exist is problematic and indicates something is very wrong.
The long startup Check test is the only test where we expect a delay (at most 30 seconds) for the files to be created.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
+        long ready0 = HealthFileUtils.getLastModifiedTime(HealthFileUtils.getReadyFile(serverRootDirFile));
+        long live0 = HealthFileUtils.getLastModifiedTime(HealthFileUtils.getLiveFile(serverRootDirFile));
+
+        Assert.assertTrue("ready should update shortly",
```
The checkInterval is at 5 seconds.
We shouldn't be waiting up to 20 seconds for an update. That means functionality is broken.
The pre-existing check is already left behind ( lines 117-120), which would be redundant.

Overall tests, could be enhanced for better tests by evaluating the timing of trace/logs that indicate when file is updated.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
-                          HealthFileUtils.isLastModifiedTimeWithinLast(HealthFileUtils.getReadyFile(serverRootDirFile), Duration.ofSeconds(12)));
-        Assert.assertTrue(Constants.LIVE_SHOULD_HAVE_UPDATED,
-                          HealthFileUtils.isLastModifiedTimeWithinLast(HealthFileUtils.getLiveFile(serverRootDirFile), Duration.ofSeconds(12)));
+        Assert.assertTrue(Constants.STARTED_SHOULD_NOT_HAVE_CREATED,
```
waitForFileMissing only checks until the file is missing. Meaning this passes immediately without checking that files aren't created after the duration.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
         Assert.assertFalse(Constants.LIVE_SHOULD_NOT_HAVE_CREATED, HealthFileUtils.getLiveFile(serverRootDirFile).exists());
 
+        Assert.assertTrue(Constants.STARTED_SHOULD_NOT_HAVE_CREATED,
+                          HealthFileUtils.waitForFileMissing(HealthFileUtils.getStartFile(serverRootDirFile), 8, 200));
```
Same comment as above, this immediately says file is not created without waiting the duration.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
-        Assert.assertFalse(Constants.LIVE_SHOULD_NOT_HAVE_CREATED, HealthFileUtils.getLiveFile(serverRootDirFile).exists());
+        Assert.assertTrue(Constants.STARTED_SHOULD_NOT_HAVE_CREATED,
+                          HealthFileUtils.waitForFileMissing(HealthFileUtils.getStartFile(serverRootDirFile), 8, 200));
+        Assert.assertTrue(Constants.READY_SHOULD_NOT_HAVE_CREATED,
```
Same as above.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
-        Assert.assertTrue(Constants.LIVE_SHOULD_HAVE_UPDATED, HealthFileUtils.isLastModifiedTimeWithinLast(HealthFileUtils.getLiveFile(serverRootDirFile), Duration.ofSeconds(8)));
-        Assert.assertFalse(Constants.READY_SHOULD_NOT_HAVE_UPDATED,
-                           HealthFileUtils.isLastModifiedTimeWithinLast(HealthFileUtils.getReadyFile(serverRootDirFile), Duration.ofSeconds(8)));
+        Assert.assertTrue("live should update quickly after toggle",
```
The checkInterval is at 5s. The liveness and readiness files are checked and updated at these intervals.
15 seconds to check for an update goes through 3 cycles. If it takes that long, then there is a problem.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.internal_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/SimpleFileBasedHealthCheckTest.java:
```diff
@@ -42,577 +40,233 @@
 
+        Assert.assertTrue("ready should update quickly after toggle",
+                          HealthFileUtils.waitForUpdateSince(readyFile, ready0, 15, 200));
+        Assert.assertTrue("live should update again",
```
15 seconds is too long. See above comment.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+        long start = System.nanoTime();
+        long timeoutNanos = Duration.ofSeconds(adjustSecondsForWindows(timeoutSeconds)).toNanos();
+        long poll = Math.max(25L, pollMillis);
+        while (System.nanoTime() - start < timeoutNanos) {
```
We don't need nanosecond precision.

Channyboy commented on dev/io.openliberty.microprofile.health.file.healthcheck.longinterval_fat/fat/src/io/openliberty/microprofile/health/file/healthcheck/fat/utils/HealthFileUtils.java:
```diff
@@ -95,4 +90,116 @@ String getFileName() {
+        long start = System.nanoTime();
+        long window = Duration.ofSeconds(adjustSecondsForWindows(windowSeconds)).toNanos();
+        long poll = Math.max(25L, pollMillis);
+        while (System.nanoTime() - start < window) {
```
Don't need nano second precision. Can avoid the perf hit.

Channyboy reviewed (commented):
Not sure if this will avoid the 3 hour time out.
Splitting tests into different classes still run it under the same FAT bundle.
Extra classes mean a little bit more overhead with the test runner setting up the classes.

Specifically for long startup check itnerval and long check interval, those tests would better benefit from checking logging/trace of the mpHealth to ensure those 30 second intervals are abided by vs timing checks that the previous code and this Prs code is trying to resolve.

Issues with the right method being used (i.e. waitForMissing). As well as unnecessary or unused code.

riyaaroraa committed:
Improve MicroProfile Health FAT tests: Replace fixed sleeps with condition-based waits and optimize timing windows

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=510f43dc-6713-490b-9f95-1cb46af7ebc2) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=b225a395-7fc4-4265-8b5d-a6ecfc9240cb) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Fix build.gradle: explicitly include fat.gradle to define autoFvtDir and other FAT variables

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6b81b4ae-6d19-4c78-9696-78d3e37df8cf) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Fix build.gradle: remove explicit fat.gradle inclusion to avoid task duplication

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=66aeda0e-e90f-4e6f-b4f1-93dfe3e8b00a) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Fix build issue by adding build.changeDetector to settings.gradle

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1437be0e-e4ba-444c-bb33-eebae5f7a437) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
longinterval_fat: apply ../wlp-gradle/fat.gradle and wire addRequiredLibraries for PB

riyaaroraa committed:
longinterval_fat: remove stray bnd.bnd

riyaaroraa committed:
longinterval_fat: use ../wlp-gradle/fat.gradle and wire addRequiredLibraries; remove extras

riyaaroraa committed:
Add missing bnd.bnd file for longinterval_fat project

riyaaroraa committed:
Fix build configuration for longinterval_fat module

- Fixed build.gradle to properly integrate with automatic FAT build system
- Added build.changeDetector to settings.gradle for dependency resolution
- Removed manual fat.gradle application and duplicate task creation

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=099fe3cf-03f3-411a-9925-949f7ddf1c09) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

riyaaroraa committed:
Update Constants.java

riyaaroraa committed:
Update fatTestPorts.xml

riyaaroraa committed:
Update testports.properties

riyaaroraa commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=81ed1a2b-1b0b-4448-bb8e-e21ceb7d7b9c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._
